File: backend/src/test/java/lan/dk/podcastserver/service/SignatureServiceTest.java
package lan.dk.podcastserver.service;

import org.apache.commons.codec.digest.DigestUtils;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.StringReader;
import java.io.UncheckedIOException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.when;

/**
 * Created by kevin on 04/09/15 for Podcast Server
 */
@RunWith(MockitoJUnitRunner.class)
public class SignatureServiceTest {

    private @Mock UrlService urlService;
    private @InjectMocks SignatureService signatureService;

    @Test
    public void should_generate_md5_from_stream() throws IOException {
        /* Given */
        String stringStream = "azertyuiopqsdfghjklmwxcvbn";
        when(urlService.asReader(anyString())).thenReturn(new BufferedReader(new StringReader(stringStream)));

        /* When */
        String s = signatureService.generateSignatureFromURL("");

        /* Then */
        assertThat(s).isEqualTo(DigestUtils.md5Hex(stringStream));
    }
    
    @Test
    public void should_return_empty_string_if_error_during_connection() throws IOException {
        /* Given */
        doThrow(UncheckedIOException.class).when(urlService).asReader(anyString());

        /* When */
        String s = signatureService.generateSignatureFromURL("");

        /* Then */
        assertThat(s).isEqualTo("");
    }

    @Test
    public void should_generate_md5_from_text() {
        /* Given */
        String stringStream = "azertyuiopqsdfghjklmwxcvbn";
        /* When */
        String s = signatureService.generateMD5Signature(stringStream);
        /* Then */
        assertThat(s).isEqualTo(DigestUtils.md5Hex(stringStream));
    }
}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/service/SignatureServiceTest.kt
package com.github.davinkevin.podcastserver.service

import com.nhaarman.mockitokotlin2.whenever
import org.apache.commons.codec.digest.DigestUtils
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers.anyString
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.Mockito.doThrow
import org.mockito.junit.jupiter.MockitoExtension
import java.io.UncheckedIOException

@ExtendWith(MockitoExtension::class)
internal class SignatureServiceTest {

    @Mock lateinit var urlService: UrlService
    @InjectMocks lateinit var signatureService: SignatureService

    @AfterEach fun afterEach() = Mockito.reset(urlService)

    @Test
    fun `should generate md5 from stream`() {
        /* Given */
        val stringStream = "azertyuiopqsdfghjklmwxcvbn"
        whenever(urlService.asStream(anyString())).thenReturn(stringStream.byteInputStream())


        /* When */
        val s = signatureService.fromUrl("")

        /* Then */
        Assertions.assertThat(s).isEqualTo(DigestUtils.md5Hex(stringStream))
    }

    @Test
    fun `should return empty string if error during connection`() {
        /* Given */
        doThrow(UncheckedIOException::class.java).whenever(urlService).asStream(anyString())

        /* When */
        val s = signatureService.fromUrl("")

        /* Then */
        Assertions.assertThat(s).isEqualTo("")
    }

    @Test
    fun `should generate md5 from text`() {
        /* Given */
        val stringStream = "azertyuiopqsdfghjklmwxcvbn"
        /* When */
        val s = signatureService.fromText(stringStream)
        /* Then */
        Assertions.assertThat(s).isEqualTo(DigestUtils.md5Hex(stringStream))
    }
}